### ISSUE TITLE
Feature request: `sort-prop-types` should work on TS Props types

### ISSUE THREAD

**Issue by tylerlaprade** [jsx-eslint/eslint-plugin-react#3578]
### Is there an existing issue for this?

- [X] I have searched the existing issues and my issue is unique
- [X] My issue appears in the command-line and not only in the text editor

### Description Overview

It currently only works on Class components, which are not in favor in the codebases in which I work. The rule doesn't trigger any warnings on properties on a functional component.

### Expected Behavior

It should behave the same on functional components as it does on class components.

### eslint-plugin-react version

v7.32.2

### eslint version

8.40.0

### node version

v16.17.1

**Comment by kshirish** [comment]
Facing the same issue. @ljharb 

```
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const variantMap = {
  filled: {
    primary: 'ui-text-white ui-bg-primary',
    secondary: 'ui-text-white ui-bg-secondary',
  },
  outlined: {
    primary: 'ui-border ui-border-primary ui-text-primary',
    secondary: 'ui-border ui-border-secondary ui-text-secondary',
  },
  text: { primary: 'ui-text-primary', secondary: 'ui-text-secondary' },
};

const sizeMap = {
  sm: 'ui-px-2 ui-py-1 ui-text-xs',
  md: 'ui-px-3 ui-py-2 ui-text-sm',
  lg: 'ui-px-4 ui-py-3 ui-text-base',
};

const Button = ({
  variant = 'filled',
  color = 'primary',
  size = 'md',
  children,
  disabled = false,
  className,
  ...restProps
}) => {
  const variantClass = variantMap[variant][color];
  const sizeClass = sizeMap[size];

  return (
    <button
      className={classNames(
        'ui-rounded',
        variantClass,
        sizeClass,
        {
          'ui-pointer-events-none': disabled,
        },
        className,
      )}
      type="button"
      {...restProps}
    >
      {children}
    </button>
  );
};

Button.propTypes = {
  variant: PropTypes.oneOf(['filled', 'outlined', 'text']),
  color: PropTypes.oneOf(['primary', 'secondary']),
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
  disabled: PropTypes.bool,
  className: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

Button.displayName = 'Button';

export default Button;

```


Although, it is highlighting the error `Prop types declarations should be sorted alphabetically  react/sort-prop-types` but `--fix` option is not fixing it!

**Comment by kshirish** [comment]
Here is the `.eslintrc.js`

```
module.exports = {
  env: {
    browser: true,
    es2021: true,
  },
  extends: [/*'eslint:recommended'*/ 'plugin:react/recommended'],
  overrides: [],
  parserOptions: {
    ecmaVersion: 'latest',
    sourceType: 'module',
  },
  plugins: ['react'],
  rules: {
    'react/sort-prop-types': [
      2,
      {
        callbacksLast: true,
        sortShapeProp: true,
        noSortAlphabetically: false,
      },
    ],
    'react/jsx-sort-props': [
      2,
      {
        callbacksLast: true,
        shorthandFirst: false,
        shorthandLast: true,
        noSortAlphabetically: false,
        reservedFirst: true,
      },
    ],
  },
};

```

**Comment by ljharb** [comment]
@kshirish autofixing isn't ever a guarantee; if it's warning properly then there's no bug, just perhaps a lack of an enhancement.

**Comment by tylerlaprade** [comment]
@ljharb, I would have expected this to give me an error since "aaa" should be sorted before "zzz".
```
type Props = {
  zzz: string;
  aaa: string;
}

function Foo(props: Props) {
...
}
```

Config:
```
"react/sort-prop-types": "error",
```

**Comment by ljharb** [comment]
does it give that error on a class component?

In other words, is this a bug with SFCs, or a bug with TS type definitions? (which i'm not sure the rule ever supported)

**Comment by tylerlaprade** [comment]
I have never used a class component, so I'm not entirely sure of the syntax, but looking at the above comment I made this change and it does fire. However, I don't think many people define their props like that anymore, at least not on any of the projects I have worked on. I really like the rules in this plugin and if possible would enjoy using this one on TypeScript components, since all my projects use them.

<img width="523" alt="image" src="https://github.com/jsx-eslint/eslint-plugin-react/assets/5475199/e151da9c-8381-4ef5-a28e-b4bea34e2a8a">

**Comment by ljharb** [comment]
`Foo.propTypes` should fire for all kinds of components, including SFCs, but i'm talking about using a Props type specifically.

**Comment by tylerlaprade** [comment]
Yes, I would like to use Props types with this rule - is it possible to add?

**Comment by ljharb** [comment]
That seems like a reasonable enhancement to add.

**Comment by akulsr0** [comment]
Should we close this? as this was fixed with #3615

**Comment by tylerlaprade** [comment]
Oh nice! Is it available for me to test out yet? Which version number is it included in?

**Comment by ljharb** [comment]
Yes, merged with #3615. It's not yet released.

**Comment by tylerlaprade** [comment]
Thanks @ljharb! When will it be released? I see that the documentation has been updated already.

**Comment by ljharb** [comment]
The docs you should be reading are on the version tag you’re using; everywhere on github, the docs on the default branch won’t necessarily be releases.

**Comment by tylerlaprade** [comment]
I see, thanks @ljharb! What is the expected release schedule?

**Comment by ljharb** [comment]
There isn't a schedule; I'll get to it as soon as I can.

**Comment by tylerlaprade** [comment]
Checking up on the status since the fix was pushed over six months ago, but there still have been no new releases. Is there a bottleneck I can help with in the release process? Does it require some sort of back-testing?

**Comment by ljharb** [comment]
Thanks for offering :-) no bottleneck, I just like to get as many PRs into a release as possible. There's 3 open PRs now I'd like to land first.
